src/textual_serve/download_manager.py

```diff
@@ -0,0 +1,175 @@
+from __future__ import annotations
+
+import asyncio
+from dataclasses import dataclass, field
+import logging
+from typing import AsyncGenerator, TYPE_CHECKING
+
+if TYPE_CHECKING:
+    from textual_serve.app_service import AppService
+
+log = logging.getLogger("textual-serve")
+
+DOWNLOAD_TIMEOUT = 4
+DOWNLOAD_CHUNK_SIZE = 1024 * 64  # 64 KB
+
+
+@dataclass
+class Download:
+    app_service: "AppService"
+    delivery_key: str
+    file_name: str
+    open_method: str
+    mime_type: str
+    encoding: str | None = None
+    """The encoding of the content.
+    Will be None if the content is binary.
+    """
+    incoming_chunks: asyncio.Queue[bytes | None] = field(default_factory=asyncio.Queue)
+
+
+class DownloadManager:
+    """Class which manages downloads for the server.
+
+    Serves as the link between the web server and app processes during downloads.
+
+    A single server has a single download manager, which manages all downloads for all
+    running app processes.
+    """
+
+    def __init__(self):
+        self._active_downloads: dict[str, Download] = {}
+        """A dictionary of active downloads.
+
+        When a delivery key is received in a meta packet, it is added to this set.
+        When the user hits the "/download/{key}" endpoint, we ensure the key is in
+        this set and start the download by requesting chunks from the app process.
+
+        When the download is complete, the app process sends a "deliver_file_end"
+        meta packet, and we remove the key from this set.
+        """
+
+    async def create_download(
+        self,
+        *,
+        app_service: "AppService",
+        delivery_key: str,
+        file_name: str,
+        open_method: str,
+        mime_type: str,
```

Comment:
Missing args from docstring...

Reply:
I think you've misread the diff here but I could be misunderstanding 😄

Reply:
Shouldn't there be a `mime_type` and `encoding` in the docstring below?

Reply:
Ah I see, added.